azurelinuxagent/ga/exthandlers.py

```diff
@@ -548,108 +561,173 @@ def _report_error_event_and_return_false(error_message):
 
         return True
 
-    def handle_ext_handler(self, ext_handler, etag):
+    def handle_ext_handler(self, ext_handler, extension, etag):
         """
         Execute the requested command for the handler and return if success
         :param ext_handler: The ExtHandler to execute the command on
+        :param extension: The extension settings on which to run the command on
         :param etag: Current incarnation of the GoalState
         :return: True if the operation was successful, False if not
         """
-        ext_handler_i = ExtHandlerInstance(ext_handler, self.protocol)
+        ext_handler_i = ExtHandlerInstance(ext_handler, self.protocol, extension=extension)
         try:
             # Ensure the extension config was valid
             if ext_handler.is_invalid_setting:
                 raise ExtensionConfigError(ext_handler.invalid_setting_reason)
 
-            state = ext_handler.properties.state
+            handler_state = ext_handler.properties.state
 
             # The Guest Agent currently only supports 1 installed version per extension on the VM.
             # If the extension version is unregistered and the customers wants to uninstall the extension,
             # we should let it go through even if the installed version doesnt exist in Handler manifest (PIR) anymore.
             # If target state is enabled and version not found in manifest, do not process the extension.
-            if ext_handler_i.decide_version(target_state=state) is None and state == ExtensionRequestedState.Enabled:
+            if ext_handler_i.decide_version(target_state=handler_state,
+                                            extension=extension) is None and handler_state == ExtHandlerRequestedState.Enabled:
                 handler_version = ext_handler_i.ext_handler.properties.version
                 name = ext_handler_i.ext_handler.name
                 err_msg = "Unable to find version {0} in manifest for extension {1}".format(handler_version, name)
                 ext_handler_i.set_operation(WALAEventOperation.Download)
-                ext_handler_i.set_handler_status(message=ustr(err_msg), code=-1)
-                ext_handler_i.report_event(message=ustr(err_msg), is_success=False)
-                return False
-
-            ext_handler_i.logger.info("Target handler state: {0} [incarnation {1}]", state, etag)
-            if state == ExtensionRequestedState.Enabled:
-                self.handle_enable(ext_handler_i)
-            elif state == ExtensionRequestedState.Disabled:
-                self.handle_disable(ext_handler_i)
-            elif state == ExtensionRequestedState.Uninstall:
-                self.handle_uninstall(ext_handler_i)
+                raise ExtensionError(msg=err_msg)
```

Comment:
what's the difference between raising and returning False? seems like we are reporting errors differently